fix race conditions & avoid leaking goroutines

unclejack opened the pull request:
This PR makes the following changes:
- the first commit makes one gofmt fix to make gofmt happy
- the second commit adds locking for messageXid; this avoids some race conditions around the messageXid
- the third commit changes the code to avoid leaking goroutines in MessageStream & avoids races
The goroutines were leaking because there was no code to stop those goroutines before.
The races were related to reading from multiple goroutines and losing the order of the messages.

These changes have been tested with ofnet. I'll test with netplugin as well.

update:
This seems to be working fine in ofnet and in netplugin.

dseevr commented on common/header.go:
```diff
@@ -5,15 +5,19 @@ package common
-		messageXid += 1
+		messageXidMutex.Lock()
+		defer messageXidMutex.Unlock()
+		messageXid++
```
This could also use `atomic.AddInt64(&messageXid, 1)` and then there's no need for a mutex at all.  Probably not worth converting the type from `int`, though

unclejack replied:
I've made this change.

dseevr replied:
👍

dseevr commented on util/stream.go:
```diff
@@ -64,9 +67,8 @@ func NewMessageStream(conn net.Conn, parser Parser) *MessageStream {
 	go m.outbound()
 	go m.inbound()
 
-	for i := 0; i < 25; i++ {
```
Why was this starting 25 parser goroutines in the first place?

unclejack replied:
This code seems to have a lot to do with the channel based buffer pool. It wasn't a good idea in the end. The messages would lose their order.

jojimt replied:
@unclejack This will need some good amount of (stress)testing. In this path, we are less concerned about the order of messages. I would suggest leaving the parallelism in place for now and just addressing the race condition and the leak.

unclejack replied:
@jojimt: There's at least an issue with that parallelism. The first issue is that there can only be 50 unread messages buffered in the channel. This means there's no point in having 25 worker goroutines to execute this parsing. Something like this would make sense in the context of a slow code path which takes at least a few milliseconds to run. This code would be blocked on the channel send as soon as 50 messages are in the queue if the messages are produced faster than they're consumed. On the consumer side in ofnet, we have only one reader process/goroutine, not 25.

The ofnet tests pass. I've also tested this change with netplugin unit tests, integration tests and system tests. The tests haven't taken longer to run because of this change either. I can imagine this may have been a potential optimization to address some channel related latency issue in Go.

unclejack replied:
@jojimt: I've done some stress testing with Hello messages to measure the impact of these changes. The code needs about 3.5 seconds to process 5000000 messages with the function we use in ofnet to parse the messages (https://github.com/contiv/ofnet/blob/master/ofctrl/ofctrl.go#L174). Do we happen to know if we need to process that many messages in such a short period of time in production?

The old code (without any of the atomic fixes and any of these locking improvements) took about 3.2-3.4 seconds. The new code with all the fixes takes about 3.3-3.7 seconds to process 5000000 messages. That's between 640 nanoseconds and 760 nanoseconds per message.

The parser function appears to be fast enough to not matter if we process in parallel or sequentially.

Do you happen to have any ideas or any other tests in mind? Please let me know if you have any other suggestions.

I've pushed a few changes to remove the use of one channel and use sync.Pool.

unclejack commented:
@abhinandanpb @jojimt: Please take a look at this PR when you have the time. Do you think we'll run into any issues if we don't process those messages in 25 goroutines in parallel?
The code should be more reliable now as the mesages will be delivered as FIFO, not randomly based on the processing order of the goroutines.

rhim commented on util/stream.go:
```diff
@@ -144,16 +147,20 @@ func (m *MessageStream) inbound() {
-		}
+		select {
+		case b := <-m.pool.Full:
+			log.Debugf("Rcvd: %v", b.Bytes())
```
Is this on the data path?

unclejack replied:
@rhim: You're right. This was logging the whole content of the received OpenFlow message. I've modified this to happen only when an error is encountered during parsing.

My initial goal was to change only what was required to solve these problems. I intend to do more refactoring later.

jojimt commented on common/header.go:
```diff
@@ -13,8 +14,9 @@ var messageXid uint32 = 1
-		messageXid += 1
-		p := Header{uint8(ver), 0, 8, messageXid}
+		atomic.AddUint32(&messageXid, 1)
+		xid := atomic.LoadUint32(&messageXid)
```
I believe the correct way to do this atomically is:
`xid := atomic.AddUint32(&messageXid, 1)`

unclejack replied:
This is fixed now.

unclejack commented:
I've validated these changes. All the netplugin tests pass with this PR.

unclejack commented:
@jojimt: Can you take a look at this PR when you have the time, please?

jojimt commented:
I think we should fix bugs rather than make design changes here.

unclejack commented:
The design was an issue on its own. I've done such work in the past. If such fixes are undesirable, it's enough to state so.

unclejack commented on util/stream.go:
```diff
@@ -81,6 +103,9 @@ func (m *MessageStream) outbound() {
 		case <-m.Shutdown:
 			log.Infof("Closing OpenFlow message stream.")
 			m.conn.Close()
+			for i := 0; i < numParserGoroutines; i++ {
```
The parser goroutines are shut down here.

unclejack commented on util/stream.go:
```diff
@@ -63,10 +85,10 @@ func NewMessageStream(conn net.Conn, parser Parser) *MessageStream {
 	go m.inbound()
-
-	for i := 0; i < 25; i++ {
+	for i := 0; i < numParserGoroutines; i++ {
```
The parser goroutines are started here. I've added a constant to make sure we clean up as many goroutines as we create.

unclejack commented:
@jojimt: I've added back those goroutines for the parser. I've commented to make it easier to see the changes made from the previous version. Those are the only changes I've made.

PTAL

unclejack commented:
I've validated these changes with ofnet. Please let me know if there's anything else I should do to get this merged.

rhim commented on util/stream.go:
```diff
@@ -52,8 +73,9 @@ type MessageStream struct {
 	m := &MessageStream{
 		conn,
-		NewBufferPool(),
+		make(chan *bytes.Buffer, 50),
```
Change 50 to a named constant as well?

unclejack replied:
@rhim: I've made the change.

unclejack commented:
@jojimt: I've removed the buffer refactor commit. The commits are also modified to avoid introducing locking for the header XID in one commit and switching that to atomic ops in another.

PTAL

dseevr commented on stream_test.go:
```diff
@@ -0,0 +1,106 @@
+		_ = <-stream.Error
+	}()
+	for i := 0; i < 5000000; i++ {
+		//fmt.Printf("attempting read %v\n", i)
```
Remove commented line here

unclejack replied:
Thanks! It's done. PTAL

dseevr commented on stream_test.go:
```diff
@@ -0,0 +1,106 @@
+	}
+	f.count++
+	copy(b, binaryMessage)
+	//fmt.Printf("written: %v bytes: %v, i: %v\n", len(data), string(data), f.i)
```
and here

unclejack commented on util/stream.go:
```diff
@@ -143,17 +151,21 @@ func (m *MessageStream) inbound() {
+		case b := <-m.pool.Full:
+			msg, err := m.parser.Parse(b.Bytes())
+			// Log all message parsing errors.
+			if err != nil {
```
Instead of logging all messages, we log only the errors. Libraries shouldn't log on their own, though. They should just provide some means to fetch the errors through return values or on a channel, depending on the code.

unclejack commented on util/stream.go:
```diff
@@ -143,17 +151,21 @@ func (m *MessageStream) inbound() {
-		if err != nil {
-			log.Print(err)
-		}
+		select {
```
This select is needed to be able to shut down the goroutines. It has no performance impact. The goroutine which processes outgoing messages is also using a select and we haven't run into any issues because of that.

I've run the benchmarks and there was no significant performance difference between the old code and the new code.

unclejack commented on common/header.go:
```diff
@@ -5,16 +5,17 @@ package common
 	return func() Header {
-		messageXid += 1
-		p := Header{uint8(ver), 0, 8, messageXid}
+		xid := atomic.AddUint32(&messageXid, 1)
```
This change was needed to avoid potential issues with the Xid being incremented concurrently. This makes the code better. It doesn't alter behavior (beyond fixing a race) and it doesn't affect performance.

unclejack commented on stream_test.go:
```diff
@@ -0,0 +1,103 @@
+package libOpenflow
```
This test and benchmark code was used to ensure the code didn't regress somehow. The tests were written to figure out how the code behaves before and after the changes. The benchmarks were used to do the same thing for performance. The behavior of the code and its performance remain the same.

rhim commented:
Since these changes were taken in contiv/libOpenflow, let’s vendor those in contiv/netplugin.

@unclejack: are the stream tests added to CI for contiv/libOpenflow, if there is any? These are good tests to run on each checkin.

- Himanshu
     .:|:.:|:.

From: unclejack <notifications@github.com>
Reply-To: contiv/libOpenflow <reply@reply.github.com>
Date: Wednesday, June 21, 2017 at 2:16 AM
To: contiv/libOpenflow <libOpenflow@noreply.github.com>
Cc: "Himanshu Raj (himansh)" <himansh@cisco.com>, Mention <mention@noreply.github.com>
Subject: Re: [contiv/libOpenflow] fix race conditions & avoid leaking goroutines (#1)


@unclejack commented on this pull request.

________________________________

In stream_test.go<https://github.com/contiv/libOpenflow/pull/1#discussion_r123194924>:

> @@ -0,0 +1,103 @@

+package libOpenflow

This test and benchmark code was used to ensure the code didn't regress somehow. The tests were written to figure out how the code behaves before and after the changes. The benchmarks were used to do the same thing for performance. The behavior of the code and its performance remain the same.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/contiv/libOpenflow/pull/1#pullrequestreview-45372533>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMiMKSbb0enjBjpxDFHY4vO1bFSnyuyiks5sGN9bgaJpZM4MOdoj>.

unclejack commented:
@rhim: All the tests and benchmarks I've mentioned have been added by this PR.

rhim commented:
Sure, I meant to say that ensure they are part of our CI whenever a PR is raised in libOpenflow it is captured.

- Himanshu
     .:|:.:|:.

From: unclejack <notifications@github.com>
Reply-To: contiv/libOpenflow <reply@reply.github.com>
Date: Thursday, June 22, 2017 at 10:17 AM
To: contiv/libOpenflow <libOpenflow@noreply.github.com>
Cc: "Himanshu Raj (himansh)" <himansh@cisco.com>, Mention <mention@noreply.github.com>
Subject: Re: [contiv/libOpenflow] fix race conditions & avoid leaking goroutines (#1)


@rhim<https://github.com/rhim>: All the tests and benchmarks I've mentioned have been added by this PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/contiv/libOpenflow/pull/1#issuecomment-310445277>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMiMKdPf3MTrL_v0H7y7izMFBXZcuxgRks5sGqGugaJpZM4MOdoj>.
